File: apps/functions/src/jrdb_scraper/raceKeyGenerator.ts
/**
 * JRDBデータファイルURL生成ユーティリティ
 */

import { JRDBDataType, getJRDBDataTypeInfo } from './entities/jrdb'

/**
 * JRDBデータファイルURLを生成
 * @param dataType - データ種別（例: 'KYI', 'KYH', 'KYG', 'KKA'）
 * @param date - 日付（Date型）
 * @returns JRDBメンバーページのURL
 * 
 * 注意: 
 * - すべてのデータタイプ: `{dataType}{YYMMDD}.lzh` 形式（例: `KYI251102.lzh`, `KZA251108.lzh`, `MZA251108.lzh`）
 */
export function generateJRDBDataFileUrl(dataType: string, date: Date): string {
  const info = getJRDBDataTypeInfo(dataType as JRDBDataType)
  
  if (!info) throw new Error(`未定義のデータタイプ: ${dataType}. jrdbDataTypeInfo.tsに定義を追加してください。`)
  
  // すべてのデータタイプは日付を含める
  const year2Digit = String(date.getFullYear()).slice(-2)
  const month = String(date.getMonth() + 1).padStart(2, '0')
  const day = String(date.getDate()).padStart(2, '0')
  const dateStr = `${year2Digit}${month}${day}`
  const fileName = `${dataType}${dateStr}.lzh`
  
  return `${info.dataFileBaseUrl}/${fileName}`
}

File: apps/prediction/src/jrdb_scraper/race_key_generator.py
"""JRDBデータファイルURL生成ユーティリティ"""

from datetime import date

from .entities.jrdb import JRDBDataType, get_jrdb_data_type_info


def generate_jrdb_data_file_url(dataType: str, date_obj: date) -> str:
    """JRDBデータファイルURLを生成
    
    Args:
        dataType: データ種別（例: 'KYI', 'KYH', 'KYG', 'KKA'）
        date_obj: 日付（date型）
    
    Returns:
        JRDBメンバーページのURL
    
    Raises:
        ValueError: 未定義のデータタイプの場合
    
    Note:
        すべてのデータタイプ: `{dataType}{YYMMDD}.lzh` 形式（例: `KYI251102.lzh`, `KZA251108.lzh`, `MZA251108.lzh`）
    """
    try:
        enumDataType = JRDBDataType(dataType.upper())
    except ValueError:
        raise ValueError(f'未定義のデータタイプ: {dataType}. jrdb.pyに定義を追加してください。')
    
    info = get_jrdb_data_type_info(enumDataType)
    
    # すべてのデータタイプは日付を含める
    year2Digit = str(date_obj.year)[-2:]
    month = f'{date_obj.month:02d}'
    day = f'{date_obj.day:02d}'
    dateStr = f'{year2Digit}{month}{day}'
    fileName = f'{dataType}{dateStr}.lzh'
    
    return f"{info['dataFileBaseUrl']}/{fileName}"

